### ISSUE TITLE
Errors when deferring assertions

### ISSUE THREAD

**Issue by rliebz** [gotestyourself/gotest.tools#111]
When attempting to `defer` an assert statement in certain scenarios, I get the following error message:

```
failed to find an expression on line XX in /path/to/my/file_test.go
```

The smallest example test I've been able to reproduce the error in is this:
```go
func TestDefer(t *testing.T) {
	defer assert.Check(t, false)
}
```

When asserting `true` instead of `false`, the test is successful, so my guess is the bug is limited to the error reporting logic.

This may be a separate issue, but when wrapping the assertion in an anonymous function call, the error becomes a panic:
```go
func TestDefer(t *testing.T) {
	defer func() { assert.Check(t, false) }()
}
```

With the following error, again only occurring on failed assertions:
```
--- FAIL: TestDefer (0.00s)
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range
```

**Comment by dnephin** [comment]
Thanks for the bug report!

I think the panic should be an easy fix. The code to lookup the expression must be making an assumption that isn't valid for anonymous functions.

"failed to find an expression" is a warning that can be printed when attempting to format an assertion failed message. I've used assertions in a multi-line anonymous function before, but I'm not sure if I've seen them fail or not.

**Comment by dnephin** [comment]
This is mostly fixed by #112, see the description in the PR for more details. There is one case that is unfortunately still ambiguous and will error, but it should generally be working better.
